File: framework/languages/node/e2e/PubSub/Client/Scenarios/ps-a2-topic-filter-scenario.ts
// PS-A2: Packet name으로 typed handler를 선택한다 시나리오를 검증한다.
import { randomUUID } from 'node:crypto';
import { PubSubNames } from '../../Shared/messages';
import { isEvent } from '../Support/evidence';
import { ensure } from '../Support/scenario-assert';
import { publishEvent, waitForAll } from './ps-a1-fanout-basic-delivery-scenario';

export async function runPsA2(publisher: string, subscribers: readonly string[]): Promise<void> {
  const runId = randomUUID().replaceAll('-', '');
  await publishEvent(publisher, PubSubNames.mainTopic, runId, 1, 'accepted');
  await publishEvent(publisher, PubSubNames.otherTopic, runId, 2, 'ignored');

  const snapshots = await waitForAll(subscribers, {
    containsAll: ['event|', `run=${runId}`, `topic=${PubSubNames.mainTopic}`],
    containsAllLineGroups: [['ignored|', `run=${runId}`, `topic=${PubSubNames.otherTopic}`]],
    timeoutMilliseconds: 10_000
  });

  ensure(snapshots.every((lines) => lines.some((line) => isEvent(line, runId, PubSubNames.mainTopic))), 'PS-A2 accepted topic was not recorded.');
  ensure(
    snapshots.every((lines) => lines.every((line) => !line.includes('event|') || !line.includes(`run=${runId}`) || !line.includes(`topic=${PubSubNames.otherTopic}`))),
    'PS-A2 non-interest topic was recorded as accepted.'
  );
  console.log('scenario PS-A2 passed');
}

File: framework/languages/dotnet/e2e/PubSub/Client/Scenarios/PsA2TopicFilterScenario.cs
// Verifies PS-A2 Topic Filter behavior.
using PubSub.Client.Support;
using PubSub.Shared;
using Zlink.HttpClient;

namespace PubSub.Client.Scenarios;

// PS-A2: verifies that subscribers accept the target topic and ignore non-interest topics.
internal static class PsA2TopicFilterScenario
{
    public static async Task RunAsync(ZLinkHttpClient publisher, IReadOnlyList<ZLinkHttpClient> subscribers)
    {
        var runId = Guid.NewGuid().ToString("N");

        // .NET exposes topic filtering at handler level here: one accepted topic and one ignored topic.
        await publisher.Post("/publish/event")
            .Query("topic", PubSubNames.MainTopic)
            .Query("runId", runId)
            .Query("sequence", "1")
            .Query("value", "accepted")
            .AsyncRaw();
        await publisher.Post("/publish/event")
            .Query("topic", PubSubNames.OtherTopic)
            .Query("runId", runId)
            .Query("sequence", "2")
            .Query("value", "ignored")
            .AsyncRaw();

        // Each subscriber records both the accepted event and the ignored marker for the other topic.
        var snapshots = await WaitForAllSubscribersAsync(
            subscribers,
            new EvidenceWaitReq(
                ["event|", $"run={runId}", $"topic={PubSubNames.MainTopic}"],
                [])
            {
                ContainsAllLineGroups =
                [
                    ["ignored|", $"run={runId}", $"topic={PubSubNames.OtherTopic}"]
                ]
            });

        // The non-interest topic must not be recorded as an accepted business event.
        ZlinkStreamAssert.Ensure(
            snapshots.All(lines => lines.Any(line => Evidence.IsEvent(line, runId, PubSubNames.MainTopic))),
            "PS-A2 accepted topic was not recorded.");
        ZlinkStreamAssert.Ensure(
            snapshots.All(lines => lines.All(line =>
                !line.Contains("event|", StringComparison.Ordinal)
                || !line.Contains($"run={runId}", StringComparison.Ordinal)
                || !line.Contains($"topic={PubSubNames.OtherTopic}", StringComparison.Ordinal))),
            "PS-A2 non-interest topic was recorded as accepted.");

        Console.WriteLine("scenario PS-A2 passed");
    }

    private static async Task<string[][]> WaitForAllSubscribersAsync(
        IReadOnlyList<ZLinkHttpClient> subscribers,
        EvidenceWaitReq request)
    {
        var waits = subscribers
            .Select(subscriber => subscriber.Post("/evidence/wait").Body(request).Async<string[]>().AsTask())
            .ToArray();
        var responses = await Task.WhenAll(waits);
        return responses.Select(response => response.Body).ToArray();
    }
}
